### PR TITLE
[sipify] make all protected methods slots

### LINKED ISSUE #45331: Regression: trying to run the multipart to singleparts algorithm throws a python error
### What is the bug or the crash?

Trying to execute the multipart to singleparts algorithm fails on QGIS master. This regression has to have occured in this dev. cycle.

### Steps to reproduce the issue

1. Launch QGIS
2. Open the processing toolbox panel
3. Try to launch multipart to singleparts algorihtm
4. Notice the python error

Stacktrace:
```
An error has occurred while executing Python code: 

RuntimeError: no access to protected functions or signals for objects not created from Python 
Traceback (most recent call last):
  File "/home/webmaster/apps/share/qgis/python/plugins/processing/gui/ProcessingToolbox.py", line 253, in executeAlgorithm
    in_place_input_parameter_name = alg.inputParameterName()
RuntimeError: no access to protected functions or signals for objects not created from Python


Python version: 3.9.7 (default, Sep  4 2021, 18:19:10) [GCC 10.3.0] 
QGIS version: 3.21.0-Master Master, b90e1be0cb 
```

### Versions

QGIS master

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by 3nids**
fixes #45331

### CODE REVIEW

**Comment by 3nids** [comment]
The spell test can discarded for now, I'm not in a mood of hacking sipify to support typos in headers.

**Comment by nyalldawson** [comment]
Fix looks acceptable to me. But I'm curious what the actual cause of this change was... did older sip versions silently treat public as private?

**Comment by nirvn** [comment]
What's the status on this? I'm missing my python algorithms :wink:

**Comment by nirvn** [comment]
@3nids , is there a reason why we double spell check here (i.e. source .h[eader] and the derived .sip.in files)? With source .h[eader] we have the #spellok tag to make the spellcheck happy. Could we just skip .sip.in files altogether since they are nowadays always generated from the headers?

File: scripts/spell_check/check_spelling.sh [hunk 1]
```diff
@@ -23,7 +23,7 @@
 # fi
 
 # extensions or files that should be excluded from file list if :% is appended in the spelling.dat file
-EXCLUDE_SCRIPT_LIST='(\.(xml|sip|pl|sh|badquote|cmake(\.in)?)|^(debian/copyright|cmake_templates/.*|tests/testdata/labeling/README.rst|tests/testdata/font/QGIS-Vera/COPYRIGHT.TXT|doc/NEWS\.html|debian/build/))$'
+EXCLUDE_SCRIPT_LIST='(\.(xml|sip|sip.in|pl|sh|badquote|cmake(\.in)?)|^(debian/copyright|cmake_templates/.*|tests/testdata/labeling/README.rst|tests/testdata/font/QGIS-Vera/COPYRIGHT.TXT|doc/NEWS\.html|debian/build/))$'
```
Comment (nirvn) [inline scripts/spell_check/check_spelling.sh:26]: Maybe better?

```suggestion
EXCLUDE_SCRIPT_LIST='(\.(xml|sip(\.in)?|pl|sh|badquote|cmake(\.in)?)|^(debian/copyright|cmake_templates/.*|tests/testdata/labeling/README.rst|tests/testdata/font/QGIS-Vera/COPYRIGHT.TXT|doc/NEWS\.html|debian/build/))$'
```

**Comment by nirvn** [comment]
@3nids , @nyalldawson , I've rebased this against master to fix the conflict. Let's merge this as soon as it turns green to avoid further conflicts.

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Ping @manisandro are you able to investigate this with a matter of urgency? If we can't resolve before Friday we'll need to revert all the sip cmake changes prior to release...

**Comment by jef-n** [comment]
BTW do we want python code to access protected members, although it's not in method of a derived class?  That's what #45331 is about, isn't it?  Or am I missing something?

**Comment by nyalldawson** [comment]
@jef-n 
> Arguable.

Only from a legalistic point of view. As you've pointed out, Python has no concept of these and we've been happily exposing the methods in the pyqgis docs, in python dir(object) results, in IDE autocomplete, etc for over a decade. It's not reasonable to expect that Python devs have had to check the c++ code/docs to determine if every method exposed in PyQGIS is supposed to be used or not.

We **CAN'T** release with this regression.  The consequences of knowingly breaking stable API in such a wholesale manner (especially mid-way through the 3.16 LTR release) are mind numbing... :scream:

**Comment by jef-n** [comment]
Hm, can't reproduce #45331 with the MSVC build.   `multiparttosingleparts` works for me from the toolbox.  And this in the python console:

``
QgsApplication.processingRegistry().algorithmById("native:multiparttosingleparts").inputParameterName()
'INPUT'
``

as in 3.20 and 3.16.

**Comment by jef-n** [comment]
sip6 with `--no-public-is-protected` also works on Linux.

**Comment by nirvn** [comment]
@nyalldawson , @jef-n , building now.

**Pull request merged**